39-linreg-gonum/linreg.go

```diff
@@ -0,0 +1,41 @@
+package main
+
+import (
+	"fmt"
+
+	"gonum.org/v1/gonum/floats"
+)
+
+func linearRegression(x, y []float64, alpha float64) (m, c float64) {
+	for i := 0; i < iterations; i++ {
+		dm, dc := computeGradient(x, y, m, c)
+		m += -dm * alpha
+		c += -dc * alpha
+		fmt.Printf("cost(%.2f, %.2f) = %.2f\n", m, c, computeCost(x, y, m, c))
+	}
+
+	fmt.Printf("cost(%.2f, %.2f) = %.2f\n", m, c, computeCost(x, y, m, c))
+
+	return m, c
+}
+
+func computeCost(x, y []float64, m, c float64) float64 {
+	floats.Scale(m, x)    // m * x
+	floats.AddConst(c, x) // m * x + c
+	floats.Sub(y, x)      // y - (m*x + c)
+	return floats.Dot(y, y) / float64(len(x))
+}
+
+func computeGradient(x, y []float64, m, c float64) (dm, dc float64) {
+	floats.Scale(m, x)    // m * x
```

Comment:
ALL x values after scale are 0

Reply:
if m is 0 yeah, but not otherwise

Reply:
FFS, the problem is that I'm actually changing the value of the points by using `Scale`!

Reply:
🙂

Reply:
I guess this is a good friction log for @sbinet and @kunde21

Reply:
Or, combine the two:
```
calc = make([]float64, len(y)
...
floats.AddScaledTo(calc, y, -m, x)  // calc = y + (-m * x)
floats.AddConst(-c, calc) // calc = (y + (-m*x) ) .- c

dm = -2 / float64(len(x)) * floats.Dot(x, calc)
dc = -2 / float64(len(x)) * floats.Sum( calc)
return dm, dc
```

Then, you could pass the `calc` slice into both compute functions as a workspace.